neuralprophet/time_dataset.py

```diff
@@ -17,6 +18,63 @@
 log = logging.getLogger("NP.time_dataset")
 
 
+class FastTensorDataLoader:
+    """
+    A DataLoader-like object for a set of tensors that can be much faster than
+    TensorDataset + DataLoader because dataloader grabs individual indices of
+    the dataset and calls cat (slow).
+    Source: https://discuss.pytorch.org/t/dataloader-much-slower-than-manual-batching/27014/6
+    """
+
+    def __init__(self, dataset, batch_size=32, shuffle=False):
+        """
+        Initialize a FastTensorDataLoader.
+        :param *tensors: tensors to store. Must have the same length @ dim 0.
+        :param batch_size: batch size to load.
+        :param shuffle: if True, shuffle the data *in-place* whenever an
+            iterator is created out of this object.
+        :returns: A FastTensorDataLoader.
+        """
+        self.dataset = dataset
+        self.dataset_len = len(dataset)
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+
+        # Calculate # batches
+        n_batches, remainder = divmod(self.dataset_len, self.batch_size)
+        if remainder > 0:
+            n_batches += 1
+        self.n_batches = n_batches
+
+    def __iter__(self):
+        if self.shuffle:
+            random.shuffle(self.dataset.combined_timedataset)
+        self.i = 0
+        return self
+
+    def __next__(self):
+        if self.i >= self.dataset_len:
+            raise StopIteration
+        # Stack the batch items
+        inputs = [inputs for inputs, _, _ in self.dataset[self.i : self.i + self.batch_size]]
+        inputs = {
+            k: torch.stack([dic[k] for dic in inputs])
+            if isinstance(inputs[0][k], torch.Tensor)
+            else {_k: torch.stack([dic[k][_k] for dic in inputs]) for _k in inputs[0][k]}
+            for k in inputs[0]
+        }
+        targets = torch.stack(
+            [targets for _, targets, _ in self.dataset[self.i : self.i + self.batch_size]]
+        )  # .squeeze(1)
+        meta = [meta for _, _, meta in self.dataset[self.i : self.i + self.batch_size]]
+        meta = {k: [dic[k] for dic in meta] for k in meta[0]}
+        self.i += self.batch_size
+        return (inputs, targets, meta)
+
+    def __len__(self):
+        return self.n_batches
+
```

Comment:
@ourownstory This would be an alternative implementation of a dataloader that supposingly brings a speedup in the batch loading within the epochs. Inspired by this: https://towardsdatascience.com/better-data-loading-20x-pytorch-speed-up-for-tabular-data-e264b9e34352